File: frontend/src/api/agentRoom.ts
// Client API pour l'Agent Room (§B28, `GET /api/agents/room/*` backend).
// Voir `backend/app/agent_room.py` pour la provenance de chaque champ — le
// "Live Debate" (messages) et le "Decision Details" (chaîne de décision)
// sont deux besoins de lecture distincts, deux endpoints distincts, jamais
// mélangés (même séparation que côté backend).

import { apiGet, apiPost } from "./client";
import type { OptionInstrument } from "./options";

export type AgentMessage = {
  id: string;
  agent_type: string;
  conversation_thread_id: string;
  // §D073 — valeurs réellement produites par le pipeline actuel :
  // `completed`/`rejected` uniquement. `thinking`/`failed` existent dans le
  // modèle (traitement asynchrone en théorie) mais ne sont jamais écrites
  // par ce pipeline synchrone à repli déterministe garanti — voir
  // AVANCEMENT.md. Typé `string` (pas une union stricte) pour ne jamais
  // fabriquer une valeur inattendue en cas d'évolution future du backend.
  state: string;
  content: string;
  payload: Record<string, unknown>;
  occurred_at: string;
};

export type AgentMessagesResponse = {
  messages: AgentMessage[];
};

export type DecisionChainProposal = {
  id: string;
  outcome: string;
  confidence: number | null;
  reasoning_text: string | null;
  risk_flags: unknown[];
  option_instrument: OptionInstrument | null;
  created_at: string;
};

export type DecisionChainCritique = DecisionChainProposal;

export type DecisionChainRiskDecision = {
  id: string;
  outcome: string;
  reasons: unknown[];
  adjustments: Record<string, unknown>;
  created_at: string;
};

export type DecisionChainExplanation = {
  id: string;
  outcome: string;
  novice_summary: string | null;
  expert_summary: string | null;
  created_at: string;
};

export type DecisionChainOrder = {
  id: string;
  symbol: string;
  side: string;
  asset_class: string;
  option_instrument: OptionInstrument | null;
  order_type: string;
  time_in_force: string;
  status: string;
  quantity: number | null;
  notional: number | null;
  filled_at: string | null;
  submitted_at: string | null;
};

export type DecisionChainResponse = {
  strategy_id: string;
  strategy_name: string | null;
  strategy_type_code: string | null;
  symbol: string;
  market_data_timestamp: string | null;
  // Chaque maillon est honnêtement `null` s'il n'a pas (encore) eu lieu —
  // jamais de 404 pour un état intermédiaire réel (voir docstring backend).
  proposal: DecisionChainProposal | null;
  critique: DecisionChainCritique | null;
  risk_decision: DecisionChainRiskDecision | null;
  explanation: DecisionChainExplanation | null;
  order: DecisionChainOrder | null;
};

export type AskZikoResponse = {
  answer: string;
  source: "claude" | "deterministic";
  decision_available: boolean;
  readonly: true;
};

export async function fetchAgentMessages(limit = 100): Promise<AgentMessagesResponse> {
  return apiGet<AgentMessagesResponse>(`/api/agents/room/messages?limit=${limit}`);
}

export async function fetchDecisionChain(
  strategyId: string,
  symbol: string,
  marketDataTimestamp: string,
): Promise<DecisionChainResponse> {
  const params = new URLSearchParams({
    strategy_id: strategyId,
    symbol,
    market_data_timestamp: marketDataTimestamp,
  });
  return apiGet<DecisionChainResponse>(`/api/agents/room/decision-chain?${params.toString()}`);
}

// Ask Ziko is intentionally bound to the same decision-window key as the
// Decision Details tab. The browser never sends a decision record: the API
// re-reads only the active context's persisted chain before explaining it.
export async function askZikoAboutDecision(input: {
  strategyId: string;
  symbol: string;
  marketDataTimestamp: string;
  question: string;
  locale: "en" | "fr" | "pt" | "es" | "de";
}): Promise<AskZikoResponse> {
  return apiPost<AskZikoResponse>("/api/agents/room/ask-ziko", {
    strategy_id: input.strategyId,
    symbol: input.symbol,
    market_data_timestamp: input.marketDataTimestamp,
    question: input.question,
    locale: input.locale,
  });
}

File: backend/app/schemas/agent_room.py
"""Schémas de lecture de l'Agent Room (§B28) — voir `backend/app/agent_room.py`
pour la justification complète de chaque source de données."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator
from shared.options import OptionInstrument


class AgentMessageOut(BaseModel):
    id: uuid.UUID
    agent_type: str
    conversation_thread_id: uuid.UUID
    state: str
    content: str
    payload: dict
    occurred_at: datetime

    model_config = {"from_attributes": True}


class AgentMessagesResponse(BaseModel):
    messages: list[AgentMessageOut]


class DecisionChainProposalOut(BaseModel):
    id: uuid.UUID
    outcome: str
    confidence: int | None
    reasoning_text: str | None
    risk_flags: list
    option_instrument: OptionInstrument | None
    created_at: datetime


class DecisionChainCritiqueOut(BaseModel):
    id: uuid.UUID
    outcome: str
    confidence: int | None
    reasoning_text: str | None
    risk_flags: list
    created_at: datetime


class DecisionChainRiskDecisionOut(BaseModel):
    id: uuid.UUID
    outcome: str
    reasons: list
    adjustments: dict
    created_at: datetime


class DecisionChainExplanationOut(BaseModel):
    id: uuid.UUID
    outcome: str
    novice_summary: str | None
    expert_summary: str | None
    created_at: datetime


class DecisionChainOrderOut(BaseModel):
    id: uuid.UUID
    symbol: str
    side: str
    asset_class: str
    option_instrument: OptionInstrument | None
    order_type: str
    time_in_force: str
    status: str
    quantity: float | None
    notional: float | None
    filled_at: datetime | None
    submitted_at: datetime | None


class DecisionChainResponse(BaseModel):
    strategy_id: uuid.UUID
    strategy_name: str | None
    strategy_type_code: str | None
    symbol: str
    market_data_timestamp: str | None
    proposal: DecisionChainProposalOut | None
    critique: DecisionChainCritiqueOut | None
    risk_decision: DecisionChainRiskDecisionOut | None
    explanation: DecisionChainExplanationOut | None
    order: DecisionChainOrderOut | None


class AskZikoRequest(BaseModel):
    """A question constrained to the Agent Room decision-window key."""

    strategy_id: uuid.UUID
    symbol: str = Field(min_length=1, max_length=20)
    market_data_timestamp: str = Field(min_length=1, max_length=64)
    question: str = Field(min_length=3, max_length=600)
    # The locale is presentation preference only, never an authorization
    # choice. Keeping it enumerated prevents a free-form prompt field.
    locale: Literal["en", "fr", "pt", "es", "de"] = "en"

    @field_validator("symbol", "market_data_timestamp", "question")
    @classmethod
    def _non_blank_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("must not be blank")
        return cleaned


class AskZikoResponse(BaseModel):
    answer: str
    source: Literal["claude", "deterministic"]
    decision_available: bool
    readonly: Literal[True] = True
